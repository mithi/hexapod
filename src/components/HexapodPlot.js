import React from "react"
import createPlotlyComponent from "react-plotly.js/factory"
import * as defaults from "../templates"
import { getNewPlotParams } from "../hexapod"

class HexapodPlot extends React.Component {
    cameraView = defaults.CAMERA_VIEW
    state = { ready: false }
    Plot = null

    logCameraView = relayoutData => {
        this.cameraView = relayoutData["scene.camera"]
    }

    componentDidMount() {
        import("plotly.js-gl3d-dist-min").then(Plotly => {
            this.Plot = createPlotlyComponent(Plotly)
            this.setState({ ready: true })
        })
    }

    render() {
        if (!this.state.ready) {
            return <p>Loading your cute robot...</p>
        }

        if (!this.props.hexapod) {
            return null
        }
        const [data, layout] = getNewPlotParams(this.props.hexapod, this.cameraView)

        const props = {
            data,
            layout,
            onRelayout: this.logCameraView,
            revision: this.props.revision,
            config: { displaylogo: false, responsive: true },
            style: { height: "100%", width: "100%" },
            useResizeHandler: true,
        }

        const Plot = this.Plot
        return <Plot {...props} />
    }
}

export default HexapodPlot
