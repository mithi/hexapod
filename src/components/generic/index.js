import NumberInputField from "./NumberInputField"
import { BasicButton, ResetButton, Card, ToggleSwitch, AlertBox } from "./SmallWidgets"
import { Slider, sliderList } from "./Slider"
export {
    BasicButton,
    ResetButton,
    Card,
    ToggleSwitch,
    AlertBox,
    NumberInputField,
    Slider,
    sliderList,
}
