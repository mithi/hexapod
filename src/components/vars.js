import React from "react"
import { GiCoffeeMug } from "react-icons/gi"
import { FaGithubAlt, FaTimes, FaCheck } from "react-icons/fa"
import { GrStatusGoodSmall } from "react-icons/gr"

const SECTION_NAMES = {
    dimensions: "Dimensions",
    inverseKinematics: "Inverse Kinematics",
    forwardKinematics: "Forward Kinematics",
    legPatterns: "Leg Patterns",
    landingPage: "Root",
}

const PATH_NAMES = {
    inverseKinematics: "/inverse-kinematics",
    forwardKinematics: "/forward-kinematics",
    legPatterns: "/leg-patterns",
    landingPage: "/",
}

const ANGLE_NAMES = ["alpha", "beta", "gamma"]
const DIMENSION_NAMES = ["front", "side", "middle", "coxia", "femur", "tibia"]
const LEG_NAMES = [
    "leftFront",
    "rightFront",
    "leftMiddle",
    "rightMiddle",
    "leftBack",
    "rightBack",
]

const IK_SLIDERS_LABELS = ["tx", "ty", "tz", "rx", "ry", "rz", "hipStance", "legStance"]
const RESET_LABEL = "reset"

/*************
 * RANGE PARAMS
 *************/

const rangeParams = absVal => ({ minVal: -absVal, maxVal: absVal, stepVal: 0.01 })
const RANGES = {
    30: rangeParams(30),
    45: rangeParams(45),
    60: rangeParams(60),
    90: rangeParams(90),
    180: rangeParams(180),
}

const translateInputs = { minVal: -1, maxVal: 1, stepVal: 0.01 }

const RANGE_PARAMS = {
    dimensionInputs: { minVal: 0, maxVal: Infinity, stepVal: 1 },
    tx: translateInputs,
    ty: translateInputs,
    tz: translateInputs,
    rx: RANGES[30],
    ry: RANGES[30],
    rz: RANGES[60],
    legStance: RANGES[90],
    hipStance: RANGES[60],
    alpha: RANGES[90],
    beta: RANGES[180],
    gamma: RANGES[180],
}

/*************
 * ICONS
 *************/

const ICON_COMPONENTS = {
    mug: <GiCoffeeMug className="vertical-align" />,
    circle: <GrStatusGoodSmall className="small-icon" />,
    octocat: <FaGithubAlt className="vertical-align" />,
    check: <FaCheck className="vertical-align" />,
    times: <FaTimes className="vertical-align" />,
}

/*************
 * NAVIGATION
 *************/

const PATHS = {
    inverseKinematics: {
        path: PATH_NAMES.inverseKinematics,
        description: SECTION_NAMES.inverseKinematics,
    },
    forwardKinematics: {
        path: PATH_NAMES.forwardKinematics,
        description: SECTION_NAMES.forwardKinematics,
    },
    legPatterns: {
        path: PATH_NAMES.legPatterns,
        description: SECTION_NAMES.legPatterns,
    },
    landingPage: {
        path: PATH_NAMES.landingPage,
        description: SECTION_NAMES.landingPage,
    },
}

const KOFI_LINK_PROPERTIES = {
    name: "KOFI",
    icon: ICON_COMPONENTS.mug,
    description: "Buy Mithi Ko-Fi üçµ",
    url: "https://ko-fi.com/minimithi",
}

const REPO_LINK_PROPERTIES = {
    name: "REPO",
    icon: ICON_COMPONENTS.octocat,
    description: "Source Code",
    url: "https://github.com/mithi/hexapod",
}

const PATH_LINKS = [
    PATHS.inverseKinematics,
    PATHS.forwardKinematics,
    PATHS.legPatterns,
    PATHS.landingPage,
]

const URL_LINKS = [KOFI_LINK_PROPERTIES, REPO_LINK_PROPERTIES]

/*************
 * LANDING PAGE
 *************/

const LANDING_PAGE_MESSAGE = `

# Mithi's Bare Minimum Hexapod Robot Simulator

- Solve (and visualize) [forward][1] and [inverse][2] kinematics purely on your browser!
It's a complete rewrite of the [one][3] I wrote in Python üêç.
No more server-side computations!

- Consider buying me a [couple cups of coffee üçµ üçµ üçµ][4] to motivate me
to build other robotics related visualizers. (Quadrotors?!)

- Love coding? However big or small, any contribution to [improve the source code][5] is always appreciated. üíô

## Love, Mithi

[1]: /forward-kinematics
[2]: /inverse-kinematics
[3]: https://github.com/mithi/hexapod-robot-simulator
[4]: https://ko-fi.com/minimithi
[5]: https://github.com/mithi/hexapod/blob/master/CONTRIBUTING.md

`

export {
    ICON_COMPONENTS,
    LANDING_PAGE_MESSAGE,
    SECTION_NAMES,
    ANGLE_NAMES,
    DIMENSION_NAMES,
    LEG_NAMES,
    IK_SLIDERS_LABELS,
    RESET_LABEL,
    PATHS,
    URL_LINKS,
    PATH_LINKS,
    RANGE_PARAMS,
}
