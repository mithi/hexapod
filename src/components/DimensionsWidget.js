import React, { Component } from "react"
import NumberInputField from "./generic/NumberInputField"
import { Card, ResetButton, ToggleSwitch } from "./generic/SmallWidgets"
import { DEFAULT_DIMENSIONS } from "../templates"
import { SECTION_NAMES, DIMENSION_NAMES, RANGE_PARAMS } from "./vars"

class DimensionsWidget extends Component {
    sectionName = SECTION_NAMES.dimensions
    state = { isFine: true }

    reset = () => this.props.onUpdate("dimensions", { dimensions: DEFAULT_DIMENSIONS })

    toggleMode = () => this.setState({ isFine: !this.state.isFine })

    updateFieldState = (name, value) => this.updateDimensions(name, value)

    updateDimensions = (name, value) => {
        const dimensions = { ...this.props.params.dimensions, [name]: value }
        this.props.onUpdate("dimensions", { dimensions })
    }

    get toggleSwitch() {
        const props = {
            id: "DimensionsWidgetSwitch",
            value: this.state.isFine ? "1x" : "5x",
            handleChange: this.toggleMode,
            showValue: true,
        }

        return <ToggleSwitch {...props} />
    }

    get NumberInputFields() {
        const { minVal, maxVal } = RANGE_PARAMS.dimensionInputs
        const stepVal = this.state.isFine ? 1 : 5
        const dimensions = this.props.params.dimensions

        const numberInputFields = DIMENSION_NAMES.map(name => {
            const props = {
                name,
                value: dimensions[name],
                rangeParams: { minVal, maxVal, stepVal },
                handleChange: this.updateFieldState,
            }

            return <NumberInputField {...props} key={name} />
        })

        return <div className="grid-cols-6">{numberInputFields}</div>
    }

    render = () => (
        <Card title={<h2>{this.sectionName}</h2>} other={this.toggleSwitch}>
            {this.NumberInputFields}
            <ResetButton reset={this.reset} />
        </Card>
    )
}

export default DimensionsWidget
