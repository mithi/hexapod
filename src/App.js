import React from "react"
import { BrowserRouter as Router } from "react-router-dom"
import { DEFAULT_POSE } from "./templates"
import { SECTION_NAMES } from "./components/vars"
import { Nav, NavDetailed, DimensionsWidget } from "./components"
import { updateHexapod, Page } from "./AppHelpers"
import HexapodPlot from "./components/HexapodPlot"

window.dataLayer = window.dataLayer || []
function gtag() {
    window.dataLayer.push(arguments)
}

class App extends React.Component {
    state = {
        inHexapodPage: false,
        hexapod: updateHexapod("default"),
        revision: 0,
    }

    /* * * * * * * * * * * * * *
     * Page load Callback
     * * * * * * * * * * * * * */

    onPageLoad = pageName => {
        document.title = pageName + " - Mithi's Bare Minimum Hexapod Robot Simulator"
        gtag("config", "UA-170794768-1", {
            page_path: window.location.pathname + window.location.search,
        })

        if (pageName === SECTION_NAMES.landingPage) {
            this.setState({ inHexapodPage: false })
            return
        }

        this.setState({ inHexapodPage: true })
        this.manageState("pose", { pose: DEFAULT_POSE })
    }

    /* * * * * * * * * * * * * *
     * State Management Callback
     * * * * * * * * * * * * * */

    manageState = (updateType, newParam) => {
        const hexapod = updateHexapod(updateType, newParam, this.state.hexapod)

        this.setState({
            revision: this.state.revision + 1,
            hexapod,
        })
    }
    /* * * * * * * * * * * * * *
     * Page Component Prototype
     * * * * * * * * * * * * * */

    pageComponent = Component => (
        <Component
            onMount={this.onPageLoad}
            onUpdate={this.manageState}
            params={{
                dimensions: this.state.hexapod.dimensions,
                pose: this.state.hexapod.pose,
            }}
        />
    )

    /* * * * * * * * * * * * * *
     * Layout
     * * * * * * * * * * * * * */

    render = () => (
        <Router>
            <Nav />
            <div id="main">
                <div id="sidebar">
                    <div hidden={!this.state.inHexapodPage}>
                        <DimensionsWidget
                            params={{ dimensions: this.state.hexapod.dimensions }}
                            onUpdate={this.manageState}
                        />
                    </div>
                    <Page pageComponent={this.pageComponent} />
                    {!this.state.inHexapodPage ? <NavDetailed /> : null}
                </div>
                <div id="plot" className="border" hidden={!this.state.inHexapodPage}>
                    <HexapodPlot
                        revision={this.state.revision}
                        hexapod={this.state.hexapod}
                    />
                </div>
            </div>
            {this.state.inHexapodPage ? <NavDetailed /> : null}
        </Router>
    )
}

export default App
