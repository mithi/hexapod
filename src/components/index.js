import { Nav, NavDetailed } from "./Nav"
import HexapodPlot from "./HexapodPlot"
import DimensionsWidget from "./DimensionsWidget"
export { Nav, NavDetailed, HexapodPlot, DimensionsWidget }
