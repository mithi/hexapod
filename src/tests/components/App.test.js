import React from "react"
import { render, screen, fireEvent } from "@testing-library/react"
import App from "../../App"
import { PATH_LINKS, URL_LINKS } from "../../components/vars"

/* * * *
 Navigation Footer
 * * * */

const expectToHaveNav = () => {
    const roles = ["navigation", "contentinfo"]
    roles.map(role => expect(screen.getByRole(role)).toBeInTheDocument())

    const allLinks = [...PATH_LINKS, ...URL_LINKS]
    allLinks.forEach(link => {
        const node = screen.getByRole("link", { name: link.description })
        expect(node).toBeInTheDocument()
    })
}

/* * * *
Default Dimensions Widget
 * * * */

const expectToHaveDefaultDimensionsWidget = () => {
    const heading = screen.getByRole("heading", { name: "Dimensions" })
    expect(heading).toBeInTheDocument()

    const dimensions = ["front", "middle", "side", "femur", "coxia", "tibia"]
    const attributes = [
        { key: "value", value: "100" },
        { key: "max", value: "Infinity" },
        { key: "min", value: "0" },
        { key: "step", value: "1" },
        { key: "type", value: "number" },
    ]

    dimensions.forEach(name => {
        expect(screen.getByLabelText(name)).toBeInTheDocument()
        const node = screen.getByRole("spinbutton", { name })
        attributes.forEach(attribute => {
            expect(node).toHaveAttribute(attribute.key, attribute.value)
        })
    })
}

/* * * *
Default Leg Patterns Page
 * * * */

const expectToHaveDefaultLegPatternsPage = () => {
    const heading = screen.getByRole("heading", { name: "Leg Patterns" })
    expect(heading).toBeInTheDocument()
    const sliderNames = ["alpha", "beta", "gamma"]

    sliderNames.forEach(name => {
        expect(screen.getByLabelText(name)).toBeInTheDocument()
        const node = screen.getByRole("slider", { name })
        expect(node).toHaveAttribute("value", "0")
        expect(node).toHaveAttribute("step", "0.01")
        expect(node).toHaveAttribute("type", "range")
    })
}

/* * * *
Default Inverse Kinematics Page
 * * * */

const expectToHaveDefaultInverseKinematics = () => {
    const heading = screen.getByRole("heading", { name: "Inverse Kinematics" })
    expect(heading).toBeInTheDocument()

    const attributes = [
        { key: "value", value: "0" },
        { key: "max", value: "1" },
        { key: "min", value: "-1" },
        { key: "step", value: "0.01" },
        { key: "type", value: "range" },
    ]

    const sliderTranslateNames = ["tx", "ty", "tz"]

    sliderTranslateNames.forEach(name => {
        expect(screen.getByLabelText(name)).toBeInTheDocument()
        const node = screen.getByRole("slider", { name })
        attributes.forEach(attribute => {
            expect(node).toHaveAttribute(attribute.key, attribute.value)
        })
    })

    const sliderRotateNames = ["rx", "ry", "rz", "hipStance", "legStance"]

    sliderRotateNames.forEach(name => {
        const node = screen.getByRole("slider", { name })
        expect(screen.getByLabelText(name)).toBeInTheDocument()
        expect(node).toHaveAttribute("value", "0")
        expect(node).toHaveAttribute("step", "0.01")
        expect(node).toHaveAttribute("type", "range")
    })
}

/* * * *
Default Forward Kinematics Page
 * * * */

const expectToHaveDefaultForwardKinematics = () => {
    const heading = screen.getByRole("heading", { name: "Forward Kinematics" })
    expect(heading).toBeInTheDocument()

    const angles = ["alpha", "beta", "gamma"]

    const legs = [
        "rightFront",
        "rightMiddle",
        "rightBack",
        "leftFront",
        "leftMiddle",
        "leftBack",
    ]

    for (const leg of legs) {
        for (const angle of angles) {
            const label = `${leg}-${angle}`
            const node = screen.getByRole((role, node) => {
                return role === "spinbutton" && node.getAttribute("id") === label
            })

            expect(node).toBeInTheDocument()
        }
    }
}

/* * * *
Elements all pages share
 * * * */

const expectEachPage = (
    flags = { numberOfResetButtons: 2, numberOfToggleSwitches: 1 }
) => {
    const resetButtons = screen.getAllByRole("button", { name: "reset" })
    const toggleSwitches = screen.getAllByRole("checkbox")
    expect(resetButtons).toHaveLength(flags.numberOfResetButtons)
    expect(toggleSwitches).toHaveLength(flags.numberOfToggleSwitches)
    expectToHaveDefaultDimensionsWidget()
    expectToHaveNav()
}

const click = name => fireEvent.click(screen.getByRole("link", { name }))

/* * * *
Application
 * * * */
describe("AppFragment", () => {
    test("Renders App snapshot correctly", () => {
        const { asFragment } = render(<App />)
        expect(asFragment()).toMatchSnapshot()
    })
})

describe("App", () => {
    beforeEach(() => {
        render(<App />)
    })

    test("Navigates to Leg Patterns page", () => {
        click("Leg Patterns")
        expectEachPage()
        expectToHaveDefaultLegPatternsPage()
    })

    test("Navigates to Inverse Kinematics page", () => {
        click("Inverse Kinematics")
        expectEachPage()
        expectToHaveDefaultInverseKinematics()
    })

    test("Navigates to Forward Kinematics page", () => {
        click("Forward Kinematics")
        expectEachPage({ numberOfResetButtons: 2, numberOfToggleSwitches: 2 })
        expectToHaveDefaultForwardKinematics()
    })

    test("Navigates to Landing Page", () => {
        click("Root")
        const heading = screen.getByRole("heading", {
            name: "Mithi's Bare Minimum Hexapod Robot Simulator",
        })
        expect(heading).toBeInTheDocument()

        const wrongHeadings = [
            "Dimensions",
            "Leg Patterns",
            "Inverse Kinematics",
            "Forward Kinematics",
        ]
        wrongHeadings.forEach(name =>
            expect(screen.queryByRole("heading", { name })).toBeNull()
        )

        expectToHaveNav()
    })
})
