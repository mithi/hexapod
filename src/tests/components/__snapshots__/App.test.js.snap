// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppFragment Renders App snapshot correctly 1`] = `
<DocumentFragment>
  <ul
    class="row-container no-bullet top-bar"
  >
    <li>
      <a
        class="link-icon"
        href="https://ko-fi.com/minimithi"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span>
          <svg
            class="vertical-align"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M221.97 54.406c-40.375 0-77.013 8.01-104.345 21.563C90.293 89.52 70.937 109.51 70.937 134c0 2.57.213 5.094.626 7.563l-.032 5.875c-.555 82.898 16.25 164.612 14.97 244.718-.372 23.295 14.294 43.908 38 57.594s56.797 21.5 97.125 21.5c40.357 0 71.338-8.024 92.906-22.03 21.57-14.008 33.408-34.712 33.408-56.908 0-3.17.054-6.345.125-9.53 42.143-1.114 75.316-23.45 97.125-52.563 22.235-29.684 34.027-66.306 34.687-99.626.66-33.32-9.983-66.91-39.156-80.594-14.588-6.842-30.593-6.573-49.25-1.125-6.16 1.798-12.548 4.25-19.126 7.375.07-2.93.125-5.86.125-8.75v-6.72c.33-2.214.5-4.484.5-6.78 0-24.49-19.357-44.478-46.69-58.03-27.33-13.554-63.937-21.564-104.31-21.564zm0 18.688c37.87 0 72.087 7.752 96.03 19.625 23.943 11.87 36.28 26.974 36.28 41.28 0 14.306-12.337 29.44-36.28 41.313-23.943 11.87-58.16 19.593-96.03 19.593-37.87 0-72.09-7.722-96.032-19.594-23.943-11.87-36.313-27.006-36.313-41.312 0-14.306 12.37-29.41 36.313-41.28 23.942-11.874 58.16-19.626 96.03-19.626zm-69.657 39.03c-23.19 6.707-37.907 16.684-37.907 27.814 0 20.193 48.398 36.562 108.094 36.562 59.696 0 108.063-16.37 108.063-36.563 0-10.264-12.52-19.546-32.657-26.187 5.266 4.105 8.22 8.692 8.22 13.563 0 17.186-36.765 31.124-82.126 31.124-45.36 0-82.125-13.938-82.125-31.125 0-5.518 3.788-10.693 10.438-15.187zm263.312 68.282c4.34-.023 7.825.793 10.53 2.063 10.826 5.076 18.406 22.81 17.908 48.124-.5 25.313-9.714 55.51-26.844 78.375-16.296 21.75-38.106 36.352-67.126 37.905 4.352-49.843 14.23-100.49 19.28-147.844 20.724-13.842 35.894-18.567 46.25-18.624z"
            />
          </svg>
           
        </span>
      </a>
    </li>
    <li>
      <a
        class="link-icon"
        href="https://github.com/mithi/hexapod"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span>
          <svg
            class="vertical-align"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 480 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1zM480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2zm-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3zm-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1z"
            />
          </svg>
           
        </span>
      </a>
    </li>
    <li>
      <a
        class="link-icon"
        href="/inverse-kinematics"
      >
        <span>
          <svg
            class="small-icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 12 12"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="6"
              cy="6"
              fill-rule="evenodd"
              r="5"
              stroke="#000"
            />
          </svg>
           
        </span>
      </a>
    </li>
    <li>
      <a
        class="link-icon"
        href="/forward-kinematics"
      >
        <span>
          <svg
            class="small-icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 12 12"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="6"
              cy="6"
              fill-rule="evenodd"
              r="5"
              stroke="#000"
            />
          </svg>
           
        </span>
      </a>
    </li>
    <li>
      <a
        class="link-icon"
        href="/leg-patterns"
      >
        <span>
          <svg
            class="small-icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 12 12"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="6"
              cy="6"
              fill-rule="evenodd"
              r="5"
              stroke="#000"
            />
          </svg>
           
        </span>
      </a>
    </li>
    <li>
      <a
        class="link-icon"
        href="/"
      >
        <span>
          <svg
            class="small-icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 12 12"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="6"
              cy="6"
              fill-rule="evenodd"
              r="5"
              stroke="#000"
            />
          </svg>
           
        </span>
      </a>
    </li>
  </ul>
  <div
    class="main content"
  >
    <div
      class="sidebar column-container cell"
    >
      <div
        class="hexapod-img"
      />
       
      <div
        id="landing"
      >
        <h1>
          Mithi's Bare Minimum Hexapod Robot Simulator
        </h1>
        <ul>
          <li>
            <p>
              Solve (and visualize) 
              <a
                href="/forward-kinematics"
              >
                forward
              </a>
               and 
              <a
                href="/inverse-kinematics"
              >
                inverse
              </a>
               kinematics purely on your browser!
It's a complete rewrite of the 
              <a
                href="https://github.com/mithi/hexapod-robot-simulator"
              >
                one
              </a>
               I wrote in Python üêç.
No more server-side computations!
            </p>
          </li>
          <li>
            <p>
              Consider buying me a 
              <a
                href="https://ko-fi.com/minimithi"
              >
                couple cups of coffee üçµ üçµ üçµ
              </a>
               to motivate me
to build other robotics related visualizers. (Quadrotors?!)
            </p>
          </li>
          <li>
            <p>
              Any contribution to improve the source code will be extremely appreciated. üíô
Anything from fixing typographical errors to completely changing the page design ... or even rewriting
modules to follow better software practices. 
              <a
                href="https://github.com/mithi/hexapod/wiki/Types-of-(code)-Contributions"
              >
                How you can help
              </a>
              .
            </p>
          </li>
        </ul>
        <h2>
          Love, Mithi
        </h2>
      </div>
      <footer>
        <nav
          id="nav"
        >
          <ul
            class="column-container no-bullet"
          >
            <li>
              <a
                class="link-icon"
                href="https://ko-fi.com/minimithi"
                rel="noopener noreferrer"
                target="_blank"
              >
                <span>
                  <svg
                    class="vertical-align"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M221.97 54.406c-40.375 0-77.013 8.01-104.345 21.563C90.293 89.52 70.937 109.51 70.937 134c0 2.57.213 5.094.626 7.563l-.032 5.875c-.555 82.898 16.25 164.612 14.97 244.718-.372 23.295 14.294 43.908 38 57.594s56.797 21.5 97.125 21.5c40.357 0 71.338-8.024 92.906-22.03 21.57-14.008 33.408-34.712 33.408-56.908 0-3.17.054-6.345.125-9.53 42.143-1.114 75.316-23.45 97.125-52.563 22.235-29.684 34.027-66.306 34.687-99.626.66-33.32-9.983-66.91-39.156-80.594-14.588-6.842-30.593-6.573-49.25-1.125-6.16 1.798-12.548 4.25-19.126 7.375.07-2.93.125-5.86.125-8.75v-6.72c.33-2.214.5-4.484.5-6.78 0-24.49-19.357-44.478-46.69-58.03-27.33-13.554-63.937-21.564-104.31-21.564zm0 18.688c37.87 0 72.087 7.752 96.03 19.625 23.943 11.87 36.28 26.974 36.28 41.28 0 14.306-12.337 29.44-36.28 41.313-23.943 11.87-58.16 19.593-96.03 19.593-37.87 0-72.09-7.722-96.032-19.594-23.943-11.87-36.313-27.006-36.313-41.312 0-14.306 12.37-29.41 36.313-41.28 23.942-11.874 58.16-19.626 96.03-19.626zm-69.657 39.03c-23.19 6.707-37.907 16.684-37.907 27.814 0 20.193 48.398 36.562 108.094 36.562 59.696 0 108.063-16.37 108.063-36.563 0-10.264-12.52-19.546-32.657-26.187 5.266 4.105 8.22 8.692 8.22 13.563 0 17.186-36.765 31.124-82.126 31.124-45.36 0-82.125-13.938-82.125-31.125 0-5.518 3.788-10.693 10.438-15.187zm263.312 68.282c4.34-.023 7.825.793 10.53 2.063 10.826 5.076 18.406 22.81 17.908 48.124-.5 25.313-9.714 55.51-26.844 78.375-16.296 21.75-38.106 36.352-67.126 37.905 4.352-49.843 14.23-100.49 19.28-147.844 20.724-13.842 35.894-18.567 46.25-18.624z"
                    />
                  </svg>
                   Buy Mithi Ko-Fi üçµ
                </span>
              </a>
            </li>
            <li>
              <a
                class="link-icon"
                href="https://github.com/mithi/hexapod"
                rel="noopener noreferrer"
                target="_blank"
              >
                <span>
                  <svg
                    class="vertical-align"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 480 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1zM480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2zm-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3zm-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1z"
                    />
                  </svg>
                   Source Code
                </span>
              </a>
            </li>
            <li>
              <a
                class="link-icon"
                href="/inverse-kinematics"
              >
                <span>
                  <svg
                    class="small-icon"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 12 12"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <circle
                      cx="6"
                      cy="6"
                      fill-rule="evenodd"
                      r="5"
                      stroke="#000"
                    />
                  </svg>
                   Inverse Kinematics
                </span>
              </a>
            </li>
            <li>
              <a
                class="link-icon"
                href="/forward-kinematics"
              >
                <span>
                  <svg
                    class="small-icon"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 12 12"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <circle
                      cx="6"
                      cy="6"
                      fill-rule="evenodd"
                      r="5"
                      stroke="#000"
                    />
                  </svg>
                   Forward Kinematics
                </span>
              </a>
            </li>
            <li>
              <a
                class="link-icon"
                href="/leg-patterns"
              >
                <span>
                  <svg
                    class="small-icon"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 12 12"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <circle
                      cx="6"
                      cy="6"
                      fill-rule="evenodd"
                      r="5"
                      stroke="#000"
                    />
                  </svg>
                   Leg Patterns
                </span>
              </a>
            </li>
            <li>
              <a
                class="link-icon"
                href="/"
              >
                <span>
                  <svg
                    class="small-icon"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 12 12"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <circle
                      cx="6"
                      cy="6"
                      fill-rule="evenodd"
                      r="5"
                      stroke="#000"
                    />
                  </svg>
                   Root
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </footer>
    </div>
    <div
      class="no-display"
    >
      <div
        style="height: 100%; width: 100%;"
      />
    </div>
  </div>
</DocumentFragment>
`;
